src/notifications/notificationmanager.cpp

```diff
@@ -563,6 +645,40 @@ void NotificationManager::removeNotificationIfUserRemovable(uint id)
     }
 }
 
+void NotificationManager::expire()
+{
+    const qint64 currentTime(QDateTime::currentDateTimeUtc().toMSecsSinceEpoch());
+    QList<uint> expiredIds;
+    qint64 nextTimeout = std::numeric_limits<qint64>::max();
+    bool unexpiredRemaining = false;
+
+    QSqlQuery expirationQuery("SELECT * FROM expiration", *database);
+    QSqlRecord expirationRecord = expirationQuery.record();
+    int expirationTableIdFieldIndex = expirationRecord.indexOf("id");
+    int expirationTableExpireAtFieldIndex = expirationRecord.indexOf("expire_at");
+    while (expirationQuery.next()) {
+        const uint id = expirationQuery.value(expirationTableIdFieldIndex).toUInt();
+        const qint64 expiry = expirationQuery.value(expirationTableExpireAtFieldIndex).value<qint64>();
+
+        if (expiry <= currentTime) {
+            expiredIds.append(id);
+        } else {
+            nextTimeout = qMin(expiry, nextTimeout);
+            unexpiredRemaining = true;
+        }
+    }
+
+    foreach (uint id, expiredIds) {
+        CloseNotification(id, NotificationExpired);
+    }
+
+    nextExpirationTime = unexpiredRemaining ? nextTimeout : 0;
+    if (nextExpirationTime) {
+        const qint64 nextTriggerInterval(nextExpirationTime - currentTime);
+        expirationTimer.start(static_cast<int>(std::min<qint64>(nextTriggerInterval, std::numeric_limits<int>::max())));
```

Comment:
nextExpirationTime must be in advance of the current time, otherwise we would have expired the notification back around line 663.